File: FP/SAST/path-traversal.py
import os
from flask import Flask, request, send_file, abort

app = Flask(__name__)

SAFE_DIR = "/var/www/html/uploads"  # Restrict files to this directory
ALLOWED_EXTENSIONS = {"txt", "jpg", "png", "pdf", "docx"}  # Allowed file types

def is_safe_path(base_path, user_input):
    abs_path = os.path.abspath(os.path.join(base_path, user_input))
    return os.path.commonpath([base_path]) == os.path.commonpath([base_path, abs_path])

@app.route("/download", methods=["GET"])
def download_file():
    filename = request.args.get("files")

    if not filename:
        return abort(400, "Missing 'files' parameter.")

    if not is_safe_path(SAFE_DIR, filename):
        return abort(403, "Access denied.")

    file_path = os.path.join(SAFE_DIR, filename)

    if "." in filename and filename.rsplit(".", 1)[1].lower() not in ALLOWED_EXTENSIONS:
        return abort(403, "File type not allowed.")

    if os.path.exists(file_path) and os.path.isfile(file_path):
        return send_file(file_path, as_attachment=True)
    else:
        return abort(404, "File not found.")

File: FP/SAST/path-traversal.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <filesystem>

bool isValidPath(const std::string& filename) {
 
    if (filename.find("..") != std::string::npos) {
        std::cerr << "Error: Invalid file path!" << std::endl;
        return false;
    }

    std::filesystem::path safeDir = "safe_files/";
    std::filesystem::path filePath = safeDir / filename;

    if (!std::filesystem::exists(filePath) || !std::filesystem::is_regular_file(filePath)) {
        std::cerr << "Error: File does not exist or is not allowed!" << std::endl;
        return false;
    }

    return true;
}

std::string readFile(const std::string& filename) {
    if (!isValidPath(filename)) {
        return "Error: Invalid request.";
    }

    std::ifstream file("safe_files/" + filename);
    if (!file) {
        return "Error: Could not open file.";
    }

    std::string contents((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    return contents;
}

int main() {
    std::string userInput;
    std::cout << "Enter the file name (inside 'safe_files/' directory): ";
    std::cin >> userInput;

    std::cout << readFile(userInput) << std::endl;
    return 0;
}
